backends/vllm/src/models/llama_nccl.rs

```diff
@@ -0,0 +1,128 @@
+use candle_core::{DType, Device, Tensor};
+use candle_nn::var_builder::ShardedSafeTensors as VarBuilder;
+use cudarc::nccl::Comm;
+use hf_hub::{api::sync::ApiBuilder, Repo, RepoType};
+use models::{llama::Config, llama_nccl::Llama, FlashAttentionMetadata};
+use std::{path::Path, rc::Rc, time::Instant};
+use tracing::info;
+
+use crate::{
+    model_executor::{
+        ModelExecutor, ModelExecutorError, ModelFilePaths, ModelLoader, ModelLoaderError,
+    },
+    models::hub_load_safetensors,
+};
+
+const LLAMA_VERSIONS_SINGLE_SAFETENSORS: &[&str] = &[
+    "TinyLlama/TinyLlama-1.1B-Chat-v1.0",
+    "meta-llama/Llama-3.2-1B",
+    "meta-llama/Llama-3.2-1B-Instruct",
+];
+
+/// Represents a Llama language model.
+///
+/// This struct encapsulates the configuration, device, data type, and the actual Llama model.
+pub struct LlamaModel {
+    /// The configuration for the Llama model.
+    config: Config,
+    /// The actual Llama model implementation.
+    model: Llama,
+}
+
+impl ModelLoader for LlamaModel {
+    type C = Config;
+    fn fetch<T: AsRef<Path>>(
+        api_key: String,
+        cache_dir: T,
+        model_id: String,
+        revision: String,
+    ) -> Result<ModelFilePaths, ModelLoaderError> {
+        let api = ApiBuilder::new()
+            .with_progress(true)
+            .with_token(Some(api_key))
+            .with_cache_dir(cache_dir.as_ref().to_path_buf())
+            .build()?;
+
+        let repo = api.repo(Repo::with_revision(
+            model_id.clone(),
+            RepoType::Model,
+            revision,
+        ));
+        let config_file_path = repo.get("config.json")?;
+        let tokenizer_file_path = repo.get("tokenizer.json")?;
+
+        let model_weights_file_paths =
+        if LLAMA_VERSIONS_SINGLE_SAFETENSORS.contains(&model_id.as_str()) {
+            vec![repo.get("model.safetensors")?]
+        } else {
+            hub_load_safetensors(&repo, "model.safetensors.index.json")?
+        };
+        
+        Ok(ModelFilePaths {
+            config_path: config_file_path,
+            tokenizer_path: tokenizer_file_path,
+            weights_path: model_weights_file_paths,
+        })
+    }
+
+    #[cfg(not(feature = "nccl"))]
```

Comment:
you don't need a feature check here, because this whole file is added only when the nccl feature is set